storage/innobase/buf/buf0flu.cc

```diff
@@ -2286,11 +2278,36 @@ static void buf_flush_page_cleaner()
 
   for (;;)
   {
+    DBUG_EXECUTE_IF("ib_page_cleaner_sleep",
+    {
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+    });
     lsn_limit= buf_flush_sync_lsn;
 
     if (UNIV_UNLIKELY(lsn_limit != 0) && UNIV_LIKELY(srv_flush_sync))
     {
     furious_flush:
+      /* During furious flush, we need to keep generating free pages. Otherwise
+      concurrent mtrs could be blocked holding latches for the pages to be
+      flushed causing deadlock in rare occasion. */
+      mysql_mutex_lock(&buf_pool.flush_list_mutex);
+      buf_pool.page_cleaner_set_idle(false);
+      bool need_eviction= buf_pool.need_LRU_eviction();
+      mysql_mutex_unlock(&buf_pool.flush_list_mutex);
+
+      if (need_eviction)
+      {
+        mysql_mutex_lock(&buf_pool.mutex);
```

Comment:
`need_eviction` is actually protected by `buf_pool.mutex`, so we’d better acquire that mutex before invoking the check. This code branch is not time critical; we can afford to widen the critical section of `buf_pool.mutex` here.

Reply:
Indeed you are right. My inference was based on the existing calls to need_LRU_eviction and it seems we never acquire buf_pool.mutex before calling it. I would refrain from modifying existing call flow to acquire stricter mutex as it could introduce performance regression. This is not an issue functionally as the atomic variable try_LRU_scan should save us.

Looks like buf_pool.flush_list_mutex is not necessary and I have removed it and added some code comments. I will also update the MDEV to capture this finding and possible future improvement.